qiskit/circuit/gate.py

```diff
@@ -42,11 +42,23 @@ def __init__(self, name: str, num_qubits: int, params: List,
     def to_matrix(self) -> np.ndarray:
         """Return a Numpy.array for the gate unitary matrix.
 
+        Returns:
+            ndarray: Matrix representation of gate.
+
         Raises:
             CircuitError: If a Gate subclass does not implement this method an
                 exception will be raised when this base class method is called.
         """
-        raise CircuitError("to_matrix not defined for this {}".format(type(self)))
+        if self.definition:
+            # Initialize an identity operator of the correct size of the circuit
+            # pylint: disable=cyclic-import
+            from qiskit.quantum_info import Operator
+            op = Operator(np.eye(2 ** self.num_qubits))
+            op._definition_to_matrix(self)
+            return op.data
```

Comment:
Couldn't we just do this? I thought the operator accepts `Gate` instances as input
```
return Operator(self).data
```

Reply:
This is doing the same thing but perhaps a little more directly. `Operator(self)` handles cases which don't seem necessary here. Also it seems more clear to me what 'definition_to_matrix' does so one doesn't need to review Operator.__init__() logic as much to see what's happening. If logic of Operator.__init__ changes it's also less likely to effect the behavior here.

Reply:
Though `definition_to_matrix` is private and potentially more volatile. We could also add a comment in the line saying `Operator(self)` takes the gate to a matrix definition, but I'm fine with the way it is right now. It just seems a little more hacky since it uses a  private method 😄

Reply:
There was some discussion in the original issue which this PR depends on. Perhaps I'll hold off here until it's resolved there.